### PR TITLE
Implementation of DeepTau SONIC producers

### PR DESCRIPTION
**Pull request by mbluj**
#### PR description:

This PR introduces the DeepTauId SONIC producer to CMSSW.
To implement the SONIC version of DeepTauId producer a new base class `DeepTauIdBase` was added. The base class contains common part of DeepTauId code while (existing) `DeepTauId` and (newly added) `DeepTauIdSONIC` derived classes provide, respectively, direct interface and interface via TritonService to a TF DeepTau model. The common code was moved from the `DeepTauId` class.

This PR requires SONIC model and configuration files to be added with https://github.com/cms-data/RecoTauTag-TrainingFiles/pull/12.

#### PR validation:

PR validated using the test script [runDeepTauIDsOnMiniAOD.py](https://github.com/cms-tau-pog/cmssw/blob/04875cec17701624459defb0dc30bdf7e9f26ef9/RecoTauTag/RecoTau/test/runDeepTauIDsOnMiniAOD.py). To enable/disable SONIC one should set the [useSONIC](https://github.com/cms-tau-pog/cmssw/blob/04875cec17701624459defb0dc30bdf7e9f26ef9/RecoTauTag/RecoTau/test/runDeepTauIDsOnMiniAOD.py#L15) flag appropriately. Tested with and without SONIC and confirmed the outputs are in both cases the same as with the unmodified code.

Tested also wf. 12434.0 (ttbar 2023 MC) which is a kind of "typical" full Run3 workflow (GEN->NanoAOD).

### CODE REVIEW

**Comment by mbluj** [comment]
FYI, @yongbinfeng

**Comment by kpedro88** [comment]
can you add the new `deepTauSonicTriton` modifier to [allSonicTriton](https://github.com/cms-sw/cmssw/blob/master/Configuration/ProcessModifiers/python/allSonicTriton_cff.py)? (apologies, didn't notice this earlier)

**Comment by kpedro88** [comment]
test parameters:
pull_requests = https://github.com/cms-data/RecoTauTag-TrainingFiles/pull/12
workflows = 11824.9001,24834.9001
relvals_opt = --what cleanedupgrade,standard,highstats,pileup,generator,extendedgen,production,identity,ged,machine,premix,nano,gpu,2017,2026

**Comment by jfernan2** [comment]
please test

**Comment by mbluj** [comment]
The matrix test issue seems unrelated to this PR, maybe restart the tests?

**Comment by kpedro88** [comment]
please test

File: RecoTauTag/RecoTau/interface/DeepTauIdBase.h [hunk 1]
```diff
@@ -0,0 +1,2192 @@
...
+#include "DataFormats/PatCandidates/interface/Tau.h"
+#include "DataFormats/TauReco/interface/TauDiscriminatorContainer.h"
+#include "DataFormats/TauReco/interface/PFTauDiscriminator.h"
+#include "DataFormats/PatCandidates/interface/PATTauDiscriminator.h"
+#include "DataFormats/Common/interface/View.h"
+#include "DataFormats/Common/interface/RefToBase.h"
+#include "DataFormats/Provenance/interface/ProductProvenance.h"
+#include "DataFormats/Provenance/interface/ProcessHistoryID.h"
+#include "FWCore/Common/interface/Provenance.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
```
Comment (kpedro88) [inline RecoTauTag/RecoTau/interface/DeepTauIdBase.h:22]: the header consistency check implies the following should be added here:
```
#include "FWCore/Framework/interface/Event.h"
```

**Comment by mbluj** [comment]
Are tests still running or they somehow hanged?
BTW, should we expect other comments to the code changes?

**Comment by kpedro88** [comment]
please test

**Comment by smuzaffar** [comment]
Milestone for this pull request has been moved to CMSSW_14_0_X.Please open a backport if it should also go in to CMSSW_13_3_X.

**Comment by jfernan2** [comment]
+1

**Comment by mbluj** [comment]
@yongbinfeng  Milestone of this development was changed to 14_0_X - do we need a backport to 13_3_X (and older)? I suppose that the backport is not needed, but please confirm.

**Comment by yongbinfeng** [comment]
> @yongbinfeng Milestone of this development was changed to 14_0_X - do we need a backport to 13_3_X (and older)? I suppose that the backport is not needed, but please confirm.

I confirm there is no need for backport.

**Comment by antoniovilela** [comment]
@mbluj @cms-sw/reconstruction-l2 
Hi guys, could we make the commit history & messages more meaningful by squashing some or all commits? Thanks.

**Comment by mbluj** [comment]
> @mbluj @cms-sw/reconstruction-l2 Hi guys, could we make the commit history & messages more meaningful by squashing some or all commits? Thanks.

@antoniovilela I do not think that squashing makes history more meaningful especially if commit descriptions are meaningful. However, I tend to agree that in this PR there are commits for small changes and I can perform some squashing to reduce their number (if really needed).

**Comment by antoniovilela** [comment]
> > @mbluj @cms-sw/reconstruction-l2 Hi guys, could we make the commit history & messages more meaningful by squashing some or all commits? Thanks.
> 
> @antoniovilela I do not think that squashing makes history more meaningful especially if commit descriptions are meaningful. However, I tend to agree that in this PR there are commits for small changes and I can perform some squashing to reduce their number (if really needed).

If commit messages are meaningful and each commit deals with separate issues, yes. If they are not, then we are better off squashing.
Thanks.

**Comment by mbluj** [comment]
@antoniovilela 11 original commits were squashed to 3, I hope it is fine.

**Comment by kpedro88** [comment]
please test

**Comment by antoniovilela** [comment]
> @antoniovilela 11 original commits were squashed to 3, I hope it is fine.

Many thanks.

**Comment by antoniovilela** [comment]
+1

**Comment by antoniovilela** [comment]
merge
   - Merging after squash.

**Comment by jfernan2** [comment]
+reconstruction

**Pull request merged**